keepalivedcp/loadbalancer.go

```diff
@@ -132,33 +137,53 @@ func (k *KeepalivedLoadBalancer) syncLoadBalancer(service *v1.Service) (*v1.Load
 				break
 			}
 
+			if annotationForwardMethod, ok := service.Annotations[serviceForwardMethodAnnotationKey]; ok {
+				forwardMethod = annotationForwardMethod
+			}
+			if forwardMethod != svc.ForwardMethod {
+				reallocateIP = false
+				break
+			}
+
 			return &v1.LoadBalancerStatus{
 				Ingress: []v1.LoadBalancerIngress{{IP: svc.IP}},
 			}, nil
 		}
 	}
 
-	var ip string
+	ip := svc.IP
 	if lbip := service.Spec.LoadBalancerIP; lbip != "" {
 		if i := net.ParseIP(lbip); i == nil {
 			return nil, fmt.Errorf("invalid loadBalancerIP specified '%s': %s", lbip, err.Error())
 		}
 		ip = lbip
-	} else {
+	} else if reallocateIP {
 		ip, err = cfg.allocateIP(k.serviceCidr)
 		if err != nil {
 			return nil, err
 		}
 	}
 
-	cfg.ensureService(serviceConfig{UID: string(service.UID), IP: ip})
+	sc := serviceConfig{
+		UID:              string(service.UID),
+		IP:               ip,
+		ServiceNamespace: service.Namespace,
+		ServiceName:      service.Name,
+		ForwardMethod:    forwardMethod,
+	}
+	cfg.ensureService(sc)
 	cfgBytes, err := cfg.encode()
 
 	if err != nil {
 		return nil, fmt.Errorf("error encoding updated config: %s", err.Error())
 	}
 
-	cm.Data[ip] = service.Namespace + "/" + service.Name
+	cm.Data = cfg.toConfigMapData()
```

Comment:
By setting `cm.Data` to the config annotation rendered to a `map[string]string`, we no longer allow users to specify their own entries in the `kube-keepalived` configmap (as they will be overwritten by `keepalived-cloud-provider`). Is this desired? Allowing manual edits in itself may be potentially risky, but I'm not sure we should explicitly disallow it.

Reply:
I think it probably makes the code a lot simpler if the provider is considered to own the config map.  It's always possible to run two instances of the keepalived-vip controller.  That's what I did while I was working on this change.